conf/defaults.yaml

```diff
@@ -231,20 +234,35 @@ topology.multilang.serializer: "org.apache.storm.multilang.JsonSerializer"
 topology.shellbolt.max.pending: 100
 topology.skip.missing.kryo.registrations: false
 topology.max.task.parallelism: null
-topology.max.spout.pending: null
+topology.max.spout.pending: 995  # Used in ACK mode only. Should be less than `topology.executor.receive.buffer.size` to avoid deadlocked cycles involving ACKer
```

Comment:
Do we need this since we addressed the netty unbounded buffering? Its a pain to tune multiple parameters to achieve back-pressure and right now max.spout.pending also needs to be tuned to get application logic like windowing to work correctly. IMO, we can try to get rid of this config if possible.

Reply:
I agree and yes I am still looking to avoid relying on this. 
Unfortunately discovered a corner case during testing that requires it as of now ... which i alluded to in that comment.  

**Details:** The issue is that ACKer bolt creates a cycle in the topology. So if ACKer bolt's recvQ is full, and also the spout's recvQ is full. Then spout & ACKer can end up in a deadlock while trying to send msg to each other. 

Setting m.s.p to a value slightly smaller than the recvQ size avoids this situation. 

One option is to make this an internal setting that is automatically set based on the revQ size and users dont have to set. 
I am investigating an alternative that allows the spout to process incoming ACKs if outbound Q is full...and thus break the deadlock. 

Hope to revert that setting back to null soon. 
I do not plan to remove that setting as now. Will remove it after ensuring @revans2's prior noted concerns in relation to removing it have been addressed.

Reply:
We may want to flow ACK tuples and metrics tuples even backpressure is now in place, given that we treat them as the thing should be handled ASAP. Given that we have tuple timeout, delaying ACK tuples could make tuple being failed and retry even tuple tree is actually completed in worst case. I myself is also not 100% sure, and that's a hard thing to verify and consider so OK to file a follow-up issue if we agree on such issue.